[AMD] Enable async pingpong for F16 GEMMs

raikonenfnu opened the pull request:
This main motivation behind this commit is to add support for PingPong with AsyncOps for GEMMs on FP16. In order to accomplish that we made these changes:
- Fork "determineDotMemoryOps" to "determineDotAsyncMemoryOps" that handles async memory ops.
- Refactor validation and pruning of memory ops to "pruneDotMemoryOps" S.T we can have clean interface for it's async memory ops counterpart "pruneAsyncDotMemoryOps".
- Add a reordering controlled by `TRITON_HIP_ENABLE_F16_ASYNC_PINGPONG` on StreamPipeliner S.T it can adjust AsyncWait stage/cluster to hoist first AsyncWait and allow set AsyncWait towards the end of the loop to make it easier for 4 PP cluster to move it before the 3rd dot-slice / 2 s_barrier before localLoads this is to ensure no race conditions.
- Modified 4 PP cluster to handle asyncOps 
- Add check to enable handling of dotSOps (dot scaled) VS dotOps (dot)
- Move `comesFromAsyncWait` to it's own AsyncUtility S.T we can reuse it on AliasScope analysis. 
- Add support for LLVM::BrOp and UnrealizedConversionCastOp for `comesFromAsyncWait` to handle AliasScope

raikonenfnu committed:
[AMD] Generalize PingPong to have different type of Load/Store Ops

This main motivation behind this commit is to add support for PingPong
with AsyncOps. In order to accomplish that we made these changes:
- Fork "determineDotMemoryOps" to "determineDotAsyncMemoryOps" that handles async memory ops.
- Refactor validation and pruning of memory ops to "pruneDotMemoryOps"
  S.T we can have clean interface for it's async memory ops counterpart
  "pruneAsyncDotMemoryOps".
- Plumb "useBlockPingpong" into StreamPipeliner S.T it can adjust AsyncWait
  stage/cluster to hoist first AsyncWait and allow set AsyncWait towards
  the end of the loop to make it easier for 4 PP cluster to move it
  before the 3rd dot-slice / 2 s_barrier before localLoads
  this is to ensure no race conditions.
- Add check to enable handling of dotSOps (dot scaled) VS dotOps (dot)

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -934,11 +1172,17 @@ void Pingponger::getDotPingponged() {
 
   //FIXME: get better condition to enable pingpong either for dot or for dot_scaled
-  if ((dotSOps.size() != 1) && (gLoadOps.size() < 2 || lLoadOps.size() < 2 || dotSOps.size() != 1)){
+  if (dotSOps.size() > 0 && dotOps.size() > 0) {
```
(dotSOps.size() > 0 || dotOps.size() > 0) ?

raikonenfnu replied:
Since this test is checking for failur case where we have both dotScaledOps and dotOps, shouldn't be `&&`, since we only want one of dotOp or dotSOp to be > 0?

jungpark-mlir replied:
Actually, !(dotSOps.size() > 0 || dotOps.size() > 0) && (dotSOps.size() > 0 && dotOps.size() > 0)
want to skip where no dot or dotS present

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -961,14 +1205,61 @@ void Pingponger::getDotPingponged() {
       return;
     }
     addAsymmetricSyncToLoop(builder, loc);
+    return;
```
thanks :)

raikonenfnu replied:
anytime :)

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -961,14 +1205,61 @@ void Pingponger::getDotPingponged() {
+      LDBG("Currently only support num_warp=8 for async PP");
+      return;
+    }
+    auto dotType = dotOps[0].getType();
```
OK for now but please comment to avoid duplicated code for tilesize later.

raikonenfnu replied:
sg!

raikonenfnu replied:
done, I moved it up to avoid duplicate of tilesize compute

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/StreamPipeline.cpp:
```diff
@@ -220,6 +225,15 @@ LogicalResult StreamPipeliner::initSchedule(int maxIndirectionLevel) {
+  // This is beneficial for maximizing hiding of latency, while ensuring
+  // 2 barriers between asyncWait and localLoad at start of loop S.T
+  // we do not hit race conditions between warp-lo and warp-hi.
+  if (useAsyncCopy && useBlockPingpong) {
```
Actually I might need this for mxfp4 gemm pingpong, need to check if this is the same change I need.

raikonenfnu replied:
sgsg!

jungpark-mlir reviewed (commented):
Basically looks good but let more people review. Especially I'm not familiar with the pipeliner code

AlexAUT commented on third_party/amd/lib/TritonAMDGPUTransforms/StreamPipeline.cpp:
```diff
@@ -220,6 +225,15 @@ LogicalResult StreamPipeliner::initSchedule(int maxIndirectionLevel) {
+  // 2 barriers between asyncWait and localLoad at start of loop S.T
+  // we do not hit race conditions between warp-lo and warp-hi.
+  if (useAsyncCopy && useBlockPingpong) {
+    stages[SCHED_ASYNC_WAIT] = stages[SCHED_LOCAL_STORE];
```
I think this should be `.. = max(0, stages[SCHED_LOCAL_LOAD] - 1)` to work correctly with num_stages  > 2. We always want the AsyncWait one stage before the LocalLoads.

raikonenfnu replied:
ah nice! will do that :)

raikonenfnu replied:
done!

AlexAUT commented:
Is there an easy way to detect before the pipeliner if we gonna pingpong a loop or not? Because we might create a worse schedule (e.g. not merging redundant AsyncWaits) if we later bail out of pingpong'ing the loop.

jungpark-mlir commented:
> Is there an easy way to detect before the pipeliner if we gonna pingpong a loop or not? Because we might create a worse schedule (e.g. not merging redundant AsyncWaits) if we later bail out of pingpong'ing the loop.

Good point, we need to refactor the passes either let pingpong pass transform the pipelining or make decision in the pipeliner.
We might defer this for now since we can still control via env var in the app level.?

raikonenfnu commented:
> Is there an easy way to detect before the pipeliner if we gonna pingpong a loop or not? Because we might create a worse schedule (e.g. not merging redundant AsyncWaits) if we later bail out of pingpong'ing the loop.

What do you guys think if for now we merge the redundant asyncCopies in blockPingPong if it fails?

raikonenfnu committed:
address comments

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu commented:
> Also have you ran the lit test?

I ran this manually and it passes, will try to run with https://github.com/triton-lang/triton/blob/0bb6019bf26954e7b5d423938e78f5a3eba9379f/test/TritonGPU/amd/amd-stream-prefetch.mlir
```
triton/build/cmake.linux-x86_64-cpython-3.10/bin/triton-opt ~/triton/test/TritonGPU/amd/amd-block-pingpong.mlir  -split-input-file --tritonamdgpu-block-pingpong="num-stages=2" | FileCheck ~/triton/test/TritonGPU/amd/amd-block-pinpong.mlir
```
Is there a ctest command that allow me to run this automatically?(without me manually doing triton-opt, filecheck on cli 😆 )

raikonenfnu commented:
> Also have you ran the lit test?

Also tested against 
```
 ~/triton/build/cmake.linux-x86_64-cpython-3.10/bin/triton-opt  ~/triton/test/TritonGPU/amd/amd-stream-prefetch.mlir -tritonamdgpu-stream-pipeline="num_stages=3 global_prefetch=1" -canonicalize | FileCheck  ~/triton/test/TritonGPU/amd/amd-stream-prefetch.mlir --check-prefixes=GLOBAL_1
 ~/triton/build/cmake.linux-x86_64-cpython-3.10/bin/triton-opt  ~/triton/test/TritonGPU/amd/amd-stream-prefetch.mlir -tritonamdgpu-stream-pipeline="num_stages=4 global_prefetch=2" -canonicalize | FileCheck  ~/triton/test/TritonGPU/amd/amd-stream-prefetch.mlir --check-prefixes=GLOBAL_2
 ~/triton/build/cmake.linux-x86_64-cpython-3.10/bin/triton-opt  ~/triton/test/TritonGPU/amd/amd-stream-prefetch.mlir -tritonamdgpu-stream-pipeline="num_stages=3 global_prefetch=1 local_prefetch=1" -canonicalize | FileCheck  ~/triton/test/TritonGPU/amd/amd-stream-prefetch.mlir --check-prefixes=GLOBAL_LOCAL_1
 ~/triton/build/cmake.linux-x86_64-cpython-3.10/bin/triton-opt  ~/triton/test/TritonGPU/amd/amd-stream-prefetch.mlir -tritonamdgpu-stream-pipeline="num_stages=2 local_prefetch=1" -canonicalize | FileCheck  ~/triton/test/TritonGPU/amd/amd-stream-prefetch.mlir --check-prefixes=LOCAL_1
```
Which also passes

raikonenfnu committed:
[AMD] Refactor comesFromAsyncWait to work on Utility.cpp/Alias

- move comesFromAsyncWait to asyncUtility to be included by
  membarUtility.cpp and utility.cpp
- Handle LLVM::BrOp and UnrealizedCastOp for it to work on aliasScope
Signed-off-by: Stanley Winata <stanley.winata@amd.com>
Co-authored-by: Alexander Weinrauch <alexander.weinrauch@amd.com>

raikonenfnu committed:
precommit lint

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

jungpark-mlir commented:
you can cd to `triton/build/cmake.linux-x86_64-cpython-3.10/` and `lit test`

raikonenfnu committed:
[AMD] Replace argument with environment to guard mxfp4 performance

Add a hipEnv to guard the streamPipeline strategy that we need for FP16
async ping-pong. Eventually we can bring back argument or something more
stable once we unify mxfp4 strategy

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu committed:
Remove unused check

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

jungpark-mlir reviewed (approved):
LGTM, just double check the skipping condition please.

raikonenfnu committed:
skip where no dot or dotS present or when is more than 1.

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu commented:
> LGTM, just double check the skipping condition please.
Actually, !(dotSOps.size() > 0 || dotOps.size() > 0) && (dotSOps.size() > 0 && dotOps.size() > 0)
want to skip where no dot or dotS present

Hey JungWook, that makes sense. I have modified the if condition to now just check that total of all dotLikeOps should be 1. I think that should handle both the cases when neither and both are present. :)

ravil-mobile commented:
Hi, @jungpark-mlir told me that we need to integrate your and my PRs. I am not sure which one should go first. My is here:  https://github.com/ROCm/triton/pull/797

I mean both of our PRs modify `async` tokens

ravil-mobile commented:
> you can cd to `triton/build/cmake.linux-x86_64-cpython-3.10/` and `lit test`

@jungpark-mlir due to changes pingpong. `TritonGPU/amd/amd-block-pingpong.mlir` fails in `shared/triton-gfx950-launch` branch
